astropy/modeling/physical_models.py

```diff
@@ -371,3 +372,334 @@ def input_units(self):
+                # Critical or Mean Overdensity Mass
+                delta = float(massfactor[1])
+                masstype = 'm'
+            else:
+                raise ValueError("Massfactor '" + str(massfactor[0]) + "' not one of 'critical', "
+                                 "'mean', or 'virial'")
+        else:
+            try:
+                # String options
+                #   virial : virial radius
+                #   Nc  : radius where density is N that of the critical density
+                #   Nm  : radius where density is N that of the mean density
+                if massfactor.lower() == "virial":
+                    # Virial Mass
+                    delta = None
+                    masstype = massfactor.lower()
+                elif massfactor[-1].lower() == 'c' or massfactor[-1].lower() == 'm':
+                    # Critical or Mean Overdensity Mass
+                    delta = float(massfactor[0:-1])
+                    masstype = massfactor[-1].lower()
+                else:
+                    raise ValueError("Massfactor "+str(massfactor)+" string not of the form "
+                                     "'#m', '#c', or 'virial'")
+            except (AttributeError, TypeError):
+                raise TypeError("Massfactor " + str(
+                    massfactor) + " not a tuple or string")
+
+        # Set density from masstype specification
+        if masstype == "virial":
+            Om_c = cosmo.Om(redshift) - 1.0
+            d_c = 18.0 * np.pi ** 2 + 82.0 * Om_c - 39.0 * Om_c ** 2
+            self.density_delta = d_c * cosmo.critical_density(redshift)
+        elif masstype == 'c':
+            self.density_delta = delta * cosmo.critical_density(redshift)
+        elif masstype == 'm':
+            self.density_delta = delta * cosmo.critical_density(redshift) * cosmo.Om(redshift)
+        else:
+            raise ValueError("Invalid masstype '" + str(masstype) +
+                             "'. Should be one of 'virial','c', or 'm'")
+
+        # Set scale radius and density
+        # Delta Mass is related to delta radius by
+        # M_{200}=\frac{4}{3}\pi r_{200}^3 200 \rho_{c}
+        # And delta radius is related to the NFW scale radius by
+        # c = R / r_{\\rm s}
+
+        # Establish mass units for density calculation (default solar masses)
+        if not isinstance(mass, u.Quantity):
+            in_mass = u.Quantity(mass, u.M_sun)
+        else:
+            in_mass = mass
+
+        # Calculate scale radius
+        # Set radial units to kiloparsec by default (unit will be rescaled by units of radius
+        # in evaluate)
+        self.radius_s = (((3.0 * in_mass) / (4.0 * np.pi * self.density_delta)) ** (
+                    1.0 / 3.0)) / concentration
+        self.radius_s = self.radius_s.to(u.kpc)
+
+        # Calculate scale density
+        self.density_s = in_mass / (4.0 * np.pi * self.radius_s ** 3 * self.A_NFW(concentration))
```

Comment:
Is there a reason to compute `radius_s` and `density_s` in the init function as well as in the `evaluate` function?
Isn't it sufficient to have the computation in the property?
It will make the code simpler.

Reply:
Two reasons:
1. The virial radius (however a user defines it) is needed for circular velocities, which a user may be interested in acquiring without evaluating densities at arbitrary radii. (i.e. if I remove the radius_s calculation from init, then a user would have to evaluate the profile before solving for circular velocities, when that technically isn't needed).
2. Since init has all the information to calculate the scale radius & density, calculating both within init enables a user to utilize all the NFW accessors without having to evaluate first. It is a convenience to the user.

Reply:
Private function added as requested.